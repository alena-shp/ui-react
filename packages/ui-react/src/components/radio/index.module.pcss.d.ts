export const radio: string
export const radio__description: string
export const radio__input: string
export const radio__label: string
export const radio__label_disabled: string
export const radio_description: string
export const radio_disabled: string


interface Stylesheet {
    'radio': string
    'radio__description': string
    'radio__input': string
    'radio__label': string
    'radio__label_disabled': string
    'radio_description': string
    'radio_disabled': string
}


declare const stylesheet: Stylesheet
export default stylesheet
