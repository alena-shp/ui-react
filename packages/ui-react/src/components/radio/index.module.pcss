.radio {
    display: grid;
    grid-auto-flow: column;
    gap: 4px 14px;
    align-items: flex-start;
    justify-content: flex-start;

    &_description {
        grid-auto-flow: initial;
        grid-template-areas: "a b" "a c";
    }

    @media (hover: hover) and (pointer: fine) {

        &:hover:not(&_disabled) {
            cursor: pointer;
        }
    }

    &__label {
        align-self: center;

        &_disabled {
            opacity: 0.4;
        }
    }

    &__input {
        margin: 0;
        padding: 0;
        appearance: none;
        border: 1px solid var(--color-border-basic);
        border-radius: 50%;
        width: 24px;
        height: 24px;
        box-sizing: border-box;
        transition: all 200ms linear;

        @media screen and (prefers-reduced-motion: reduce) {
            transition: none;
        }

        @media (hover: hover) and (pointer: fine) {

            &:hover:not([disabled]) {
                cursor: pointer;
                border-color: var(--color-border-brand);
            }
        }

        &:checked {
            border-color: var(--color-bg-brand-1);
            border-width: 7px;

            @media (hover: hover) and (pointer: fine) {

                &:hover:not([disabled]) {
                    border-color: var(--color-bg-brand);
                }
            }
        }

        &[disabled] {
            background-color: var(--color-bg-plate);
        }

        &:checked[disabled] {
            border-color: var(--color-bg-plate-1);
            background-color: transparent;
        }

        &:focus:not([disabled]) {
            outline: none;
            box-shadow: 0 0 0 2px rgba(80, 103, 222, 0.2);
        }
    }

    &__description {
        color: var(--color-text-secondary);
        grid-area: c;
    }
}
