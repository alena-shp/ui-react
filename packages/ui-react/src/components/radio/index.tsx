import React, { forwardRef, InputHTMLAttributes, ReactNode, useMemo } from 'react';

import { TStyle, useClassnames } from '../../hooks/use-classnames';
import { Text } from '../typography/text';

import style from './index.module.pcss';

type TNative = InputHTMLAttributes<HTMLInputElement>;

export interface IProps {
    /**
     * Название поля ввода, которое будет отправлено на сервер вместе со значением.
     **/
    name: string,
    /**
     * Позволяет добавить пользовательские CSS-классы
     **/
    className?: string | TStyle,
    /**
     * Текст метки, связанной с переключателем.
     **/
    label?: ReactNode,
    /**
     * Значение, которое будет отправлено на сервер при отправке формы.
     **/
    value?: string,
    /**
     * Значение переключателя по умолчанию.
     **/
    defaultValue?: string,
    /**
     * Дополнительный текст, который поясняет, что выбирает переключатель.
     **/
    description?: ReactNode,
    /**
     * Устанавливает или определяет, должен ли компонент быть отмечен.
     **/
    checked?: TNative['checked'],
    /**
     * Определяет, должен ли флажок быть установлен при первоначальной загрузке компонента.
     **/
    defaultChecked?: TNative['defaultChecked'],
    /**
     * Функция обратного вызова, которая будет вызываться при изменении значения переключателя.
     **/
    onChange?: TNative['onChange'],
    /**
     * Функция обратного вызова, которая будет вызываться при клике на переключатель.
     **/
    onClick?: TNative['onClick'],
    /**
     * Функция обратного вызова, которая будет вызываться при потере фокуса на переключателе.
     **/
    onBlur?: TNative['onBlur'],
    /**
     * Функция обратного вызова, которая будет вызываться при получении фокуса на переключателе.
     **/
    onFocus?: TNative['onFocus'],
    /**
     * Указывает, должен ли компонент получать фокус при монтировании (отрисовке) на странице.
     **/
    autoFocus?: TNative['autoFocus'],
    /**
     * Уникальный идентификатор переключателя.
     **/
    id?: TNative['id'],
    /**
     * Порядковый номер элемента при переключении по `Tab`.
     **/
    tabIndex?: TNative['tabIndex'],
    /**
     * Определяет, будет ли компонент отключен.
     **/
    disabled?: TNative['disabled']
}

export const Radio = forwardRef<HTMLInputElement | null, IProps>(({ tabIndex = 0, ...props }, ref) => {
    const cn = useClassnames(style, props.className);

    const elDescription = useMemo(() => {
        if(props.description) {
            return (
                <span
                    className={cn('radio__description')}
                    children={props.description}
                />
            );
        }
    }, [props.description]);

    return (
        <label
            id={props.id}
            className={cn('radio', {
                'radio_disabled'   : props.disabled,
                'radio_description': !!props.description
            })}
        >
            <input
                ref={ref}
                id={props.id ? `${props.id}-input` : undefined}
                name={props.name}
                role="radio"
                type="radio"
                aria-checked={!!props.checked}
                checked={props.checked}
                defaultChecked={props.defaultChecked}
                value={props.value}
                defaultValue={props.defaultValue}
                onChange={props.onChange}
                onBlur={props.onBlur}
                onFocus={props.onFocus}
                onClick={props.onClick}
                autoFocus={props.autoFocus}
                disabled={props.disabled}
                tabIndex={props.disabled ? -1 : tabIndex}
                className={cn('radio__input')}
            />
            <Text className={cn('radio__label', { 'radio__label_disabled': props.disabled })}>
                {props.label}
            </Text>
            {elDescription}
        </label>
    );
});
