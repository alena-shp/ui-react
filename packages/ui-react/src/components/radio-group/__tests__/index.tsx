import React from 'react';
import { create } from 'react-test-renderer';

import { RadioGroup } from '..';
import { Radio } from '../../radio';

const children = [
    <Radio
        key="1"
        name="name-1"
        label="label-1"
    />,
    <Radio
        key="2"
        name="name-2"
        label="label-2"
    />,
    <Radio
        key="3"
        name="name-3"
        label="label-3"
    />,
    <Radio
        key="4"
        name="name-4"
        label="label-4"
    />
];

test('Default', () => {
    const tree = create(
        <RadioGroup
            children={children}
        />
    );

    expect(tree.toJSON()).toMatchSnapshot();
});

test('Legend', () => {
    const tree = create(
        <RadioGroup
            children={children}
            legend="Legend"
        />
    );

    expect(tree.toJSON()).toMatchSnapshot();
});

test('Legend Object', () => {
    const tree = create(
        <RadioGroup
            children={children}
            legend={{
                presetColor: 'error',
                presetSize: 'caption'
            }}
        />
    );

    expect(tree.toJSON()).toMatchSnapshot();
});

