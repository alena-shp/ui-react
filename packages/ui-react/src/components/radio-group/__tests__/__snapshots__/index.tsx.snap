// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default 1`] = `
<fieldset
  className="radio-group radio-group_row"
>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-1"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-1
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-2"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-2
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-3"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-3
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-4"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-4
    </p>
  </label>
</fieldset>
`;

exports[`Legend 1`] = `
<fieldset
  className="radio-group radio-group_row"
>
  <legend
    className="typography typography_size-body typography_color-basic"
  >
    Legend
  </legend>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-1"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-1
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-2"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-2
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-3"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-3
    </p>
  </label>
  <label
    className="radio"
  >
    <input
      aria-checked={false}
      className="radio__input"
      name="name-4"
      role="radio"
      tabIndex={0}
      type="radio"
    />
    <p
      className="typography typography_size-body typography_color-basic"
    >
      label-4
    </p>
  </label>
</fieldset>
`;
